### ISSUE TITLE
Memory leak on Table.Query only when a FilterExpression is set on QueryOperationConfig

### ISSUE THREAD

**Issue by victorvartan** [aws/aws-sdk-net#863]
The memory leak seems to be generated by LitJson. It happens only if I specify different filter expressions, if I do the query without a filter expression then it works...

Also, it happens from the first Query call, I'm not doing something recursive.

Screenshot with the memory usage from a debugging session:

![2018-02-09 3](https://user-images.githubusercontent.com/3120594/36002808-29f22e94-0d34-11e8-89fd-2cdf04bb4d0e.png)

## Expected Behavior
The memory usage should not increase

## Current Behavior
The machine's available memory shrinks fast until it's barely usable.

## Possible Solution
Use a good library for JSON de/serialization such as https://www.newtonsoft.com/json ?

## Steps to Reproduce (for bugs)
Unfortunately I don't have time now to create a test project etc. I have a lot of other tasks, maybe in a couple of weeks I'll have more free time.

## Context
The server stopped taking in other requests as a result from this.

## Your Environment

* AWSSDK.Core version used: 3.3.21.10
* Service assembly and version used: AWSSDK.DynamoDBv2 v3.3.5
* Operating System and version: Windows 10 Home, version 1709
* Visual Studio version: 15.5.2
* Targeted .NET platform: .Net Core 2.0

## .NET Core Info

* .NET Core version used for development: 2.0.5
* .NET Core version installed in the environment where application runs: 2.0.5
* Output of `dotnet --info`: 
.NET Command Line Tools (2.1.2)

Product Information:
 Version:            2.1.2
 Commit SHA-1 hash:  5695315371

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.16299
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\2.1.2\

Microsoft .NET Core Shared Framework Host

  Version  : 2.0.3
  Build    : a9190d4a75f4a982ae4b4fa8d1a24526566c69df
* Contents of project.json/project.csproj:

**Comment by sstevenkang** [comment]
This is super interesting.  I've been looking into dotnet runtime memory hogging behavior on linux past few days.  I'll take a look.

The reason we have a "copy" of LitJson is because we don't want to introduce a dependency & possible dependency collision.

**Comment by victorvartan** [comment]
TL;DR: the issue was from a custom global exception handler, not related to LitJson even though strangely it showed up in the diagnostics as having the biggest slice of memory (screenshot above)...

The memory leak happened if:
1. the query uses an IdexName in the QueryOperationConfig that doesn't exist in the table, or
2. an attribute used in the FilterExpression is in a primary key attribute for the index used in the query. This was my case, as I have a secondary index with DateCreated as the sort key and wanted to also filter on it like "#DateCreated BETWEEN :DateCreatedMin AND :DateCreatedMax"

I took some time today to build a small WebApi test project and to my surprise querying the same DynamoDB table instead of memory leaks I got Amazon.DynamoDBv2.AmazonDynamoDBException:
1. The table does not have the specified index: INVALID-index
2. Filter Expression can only contain non-primary key attributes: Primary key attribute: DateCreated

That got me wondering if maybe LitJson could have generated the memory leak while trying to de-serialize these exceptions...
Went in Startup.cs to see if there is something related that could cause this... 
app.UseMiddleware(typeof(ExceptionHandlerMiddleware)); // global exception handling
And the memory leak was from a loop in our custom exception handler...

So sorry if the report didn't find the cause of your memory leak, I hope that maybe at least by accident the problem was related to LitJson and the bug report pointed you in the right direction.
Maybe a memory profiler would help you track it down.

Also, I don't know how hard it is to implement with the current DynamoDb engine architecture, but it would be super-useful if we could use a sort key attribute in a filter expression. 
For example I think it makes sense if you have entities in an index with a sort key (like DateCreated in my case) to be able to also filter the values of that sort key before sorting them.

Thanks!
Victor
